net/mptcp/mptcp_blest.c

```diff
@@ -0,0 +1,487 @@
+
+			skb = mptcp_blest_rcv_buf_optimization(subsk, 0);
+			if (skb)
+				*reinject = -1;
+		}
+	}
+	return skb;
+}
+
+/* copy from mptcp_sched.c: mptcp_next_segment */
+static struct sk_buff *mptcp_blest_next_segment(struct sock *meta_sk,
+						int *reinject,
+						struct sock **subsk,
+						unsigned int *limit)
+{
+	struct sk_buff *skb = __mptcp_blest_next_segment(meta_sk, reinject);
+	unsigned int mss_now;
+	struct tcp_sock *subtp;
+	u16 gso_max_segs;
+	u32 max_len, max_segs, window, needed;
+
+	/* As we set it, we have to reset it as well. */
+	*limit = 0;
+
+	if (!skb)
+		return NULL;
+
+	*subsk = blest_get_available_subflow(meta_sk, skb, false);
+	if (!*subsk)
+		return NULL;
+
+	subtp = tcp_sk(*subsk);
+	mss_now = tcp_current_mss(*subsk);
+
+	if (!*reinject && unlikely(!tcp_snd_wnd_test(tcp_sk(meta_sk), skb, mss_now))) {
+		skb = mptcp_blest_rcv_buf_optimization(*subsk, 1);
+		if (skb)
+			*reinject = -1;
+		else
+			return NULL;
+	}
+
+	/* No splitting required, as we will only send one single segment */
+	if (skb->len <= mss_now)
+		return skb;
+
+	/* The following is similar to tcp_mss_split_point, but
+	 * we do not care about nagle, because we will anyways
+	 * use TCP_NAGLE_PUSH, which overrides this.
+	 *
+	 * So, we first limit according to the cwnd/gso-size and then according
+	 * to the subflow's window.
+	 */
+
+	gso_max_segs = (*subsk)->sk_gso_max_segs;
+	if (!gso_max_segs) /* No gso supported on the subflow's NIC */
+		gso_max_segs = 1;
+	max_segs = min_t(unsigned int, tcp_cwnd_test(subtp, skb), gso_max_segs);
+	if (!max_segs)
+		return NULL;
+
+	max_len = mss_now * max_segs;
+	window = tcp_wnd_end(subtp) - subtp->write_seq;
+
+	needed = min(skb->len, window);
+	if (max_len <= skb->len)
+		/* Take max_win, which is actually the cwnd/gso-size */
+		*limit = max_len;
+	else
+		/* Or, take the window */
+		*limit = needed;
+
+	return skb;
+}
+
+static void blestsched_init(struct sock *sk)
+{
+	struct blestsched_priv *bsp = blestsched_get_priv(tcp_sk(sk));
+	struct blestsched_cb *metabsp = blestsched_get_cb(tcp_sk(mptcp_meta_sk(sk)));
+
+	bsp->last_rbuf_opti = tcp_jiffies32;
+	bsp->min_srtt = 0xffffffff;
+	bsp->max_srtt = 0;
+
+	if (!metabsp->lambda_1000) {
+		metabsp->lambda_1000 = lambda * 100;
+		metabsp->last_lambda_update = tcp_jiffies32;
+	}
+}
+
+static struct mptcp_sched_ops mptcp_blest = {
+	.get_subflow = blest_get_available_subflow,
+	.next_segment = mptcp_blest_next_segment,
+	.init = blestsched_init,
+	.name = "blest",
+	.owner = THIS_MODULE,
+};
+
+static int __init blest_register(void)
+{
+	BUILD_BUG_ON(sizeof(struct blestsched_priv) > MPTCP_SCHED_SIZE);
+	BUILD_BUG_ON(sizeof(struct blestsched_cb) > MPTCP_SCHED_DATA_SIZE);
+
+	if (mptcp_register_scheduler(&mptcp_blest))
+		return -1;
+
+	return 0;
+}
+
+static void blest_unregister(void)
+{
+	mptcp_unregister_scheduler(&mptcp_blest);
+}
+
+module_init(blest_register);
+module_exit(blest_unregister);
+
+MODULE_AUTHOR("Simone Ferlin, Christoph Paasch");
```

Comment:
I don't think I should be on that list ;-)

Reply:
Ok, but you were already on the list in the version I started working on 
>https://bitbucket.org/blest_mptcp/mptcp-paper-ifip2016

I kept this line untouched and think Simone put you on the list because she reused large parts from the default scheduler implementation. I just want to give proper credit to anyone who was involved. I also added a short header to the file to mention the people that initially implement and evaluated the algorithm.

For my part, I just took Simone's implementation and used it for my master's thesis. Also, I'm not sure if I belong to this list because I only refactored BLEST to match with the rest of the MPTCP project and made no functional changes to the algorithm. This is my very first "real" contribution to the open source community. So feel free to correct me if I'm giving the credits in the wrong way.

Reply:
> Ok, but you were already on the list in the version I started working on
> 
> > https://bitbucket.org/blest_mptcp/mptcp-paper-ifip2016
> 
> I kept this line untouched and think Simone put you on the list because she reused large parts from the default scheduler implementation. I just want to give proper credit to anyone who was involved. I also added a short header to the file to mention the people that initially implement and evaluated the algorithm.
> 
> For my part, I just took Simone's implementation and used it for my master's thesis. Also, I'm not sure if I belong to this list because I only refactored BLEST to match with the rest of the MPTCP project and made no functional changes to the algorithm. This is my very first "real" contribution to the open source community. So feel free to correct me if I'm giving the credits in the wrong way.

I think that the work of the one who picks up the pieces and puts the code in a form that is "upstreamable", creates the pull-requests, spends time interacting with the maintainers, applying the feedback,... is as valuable as the one who did the initial prototype implementation.

Thus, you definitely deserve the credit !